dandischema/digests/zarr.py

```diff
@@ -0,0 +1,148 @@
+            except ValueError:
+                self.files.append(new_checksum)
+        self.files = sorted(self.files)
+
+    def add_directory_checksums(self, checksums: List[ZarrChecksum]):
+        """Add a list of directory checksums to the listing."""
+        for new_checksum in checksums:
+            try:
+                self.directories[
+                    self._index(self.directories, new_checksum)
+                ] = new_checksum
+            except ValueError:
+                self.directories.append(new_checksum)
+        self.directories = sorted(self.directories)
+
+    def remove_checksums(self, paths: List[str]):
+        """Remove a list of paths from the listing."""
+        self.files = sorted(
+            filter(lambda checksum: checksum.path not in paths, self.files)
+        )
+        self.directories = sorted(
+            filter(lambda checksum: checksum.path not in paths, self.directories)
+        )
+
+
+class ZarrChecksumListing(pydantic.BaseModel):
+    """
+    A listing of checksums for all sub-files/directories in a zarr directory.
+
+    This is the data serialized in the checksum file.
+    """
+
+    checksums: ZarrChecksums
+    md5: str
+
+
+class ZarrJSONChecksumSerializer:
+    def aggregate_checksum(self, checksums: ZarrChecksums) -> str:
+        """Generate an aggregated checksum for a list of ZarrChecksums."""
+        # Use the most compact separators possible
+        # content = json.dumps([asdict(zarr_md5) for zarr_md5 in checksums], separators=(',', ':'))0
+        content = checksums.json(**ENCODING_KWARGS)
+        h = hashlib.md5()
+        h.update(content.encode("utf-8"))
+        return h.hexdigest()
+
+    def serialize(self, zarr_checksum_listing: ZarrChecksumListing) -> str:
+        """Serialize a ZarrChecksumListing into a string."""
+        # return json.dumps(asdict(zarr_checksum_listing))
+        return zarr_checksum_listing.json(**ENCODING_KWARGS)
+
+    def deserialize(self, json_str: str) -> ZarrChecksumListing:
+        """Deserialize a string into a ZarrChecksumListing."""
+        # listing = ZarrChecksumListing(**json.loads(json_str))
+        # listing.checksums = [ZarrChecksum(**checksum) for checksum in listing.checksums]
+        # return listing
+        return ZarrChecksumListing.parse_raw(json_str)
+
+    def generate_listing(
+        self,
+        checksums: Optional[ZarrChecksums] = None,
+        files: Optional[List[ZarrChecksum]] = None,
+        directories: Optional[List[ZarrChecksum]] = None,
+    ) -> ZarrChecksumListing:
+        """
+        Generate a new ZarrChecksumListing from the given checksums.
+
+        This method wraps aggregate_checksum and should not be overridden.
+        """
+        print(files)
+        if checksums is None:
+            checksums = ZarrChecksums(
+                files=files if files is not None else [],
+                directories=directories if directories is not None else [],
```

Comment:
I think it would be safer to sort `files` and `directories` here rather than doing it in `get_checksum()` or expecting the caller to do it.